packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,276 @@
+      target.getClass()
+    else
+      target
+  )
+    let (mappedInputs =
+      if (targetClass == Dynamic)
+        applyToDynamic(List(), target)
+      else
+        applyToClass(targetClass, List()))
+    if (assertNoFailures(targetClass, mappedInputs) && targetClass != Dynamic)
+      deepToTyped.apply(targetClass, mappedInputs)
+    else
+      mappedInputs
+
+/// The scheme used to read resources from.
+///
+/// By default, this is `"prop:"` so values are read from external
+/// properties passed using the `-p` flag.
+hidden inputScheme: String(matches(Regex(#"[\w-]+:"#)))
+
+/// Path separator used to when building resource URI.
+///
+/// Default is `.` which means that nested values are passed in
+/// using a dot seperated path as the key value.
+hidden pathSeparator: String
+
+/// Template used to convert a path of elements into a resource URI.
+///
+/// Default joins the path components with [pathSeparator] and
+/// prepends the [inputScheme].
+hidden pathTemplate: (List<String>) -> String = (path: List<String>) -> "\(inputScheme)\(path.join(pathSeparator))"
+
+/// Function used to read in resources. Can be overwritten to change how
+/// values are read. Useful for testing.
+hidden readFunc = (s) -> read?(s)
+
+function applyToClass(target: Class, path: List<String>): Mapping<String, Any?> =
+  let (reflected = reflect.Class(target))
+  new Mapping {
+    for (name, property in reflected.properties) {
+      when (property.type is Coerceable) {
+        [name] = let (potentialValue = readValueToTarget(path, name, property))
+            if (!(potentialValue is ReadFailure)) potentialValue
+            else if (property.type is reflect.NullableType) property.defaultValue
+            else property.defaultValue ?? potentialValue
+      }
+    }
+  }
+
+typealias Coerceable = reflect.DeclaredType | reflect.NullableType | reflect.TypeAlias | reflect.UnknownType
+
+local function assertNoFailures(target: Class, output: Mapping<String, Any?>|Dynamic): Boolean =
+  let (fails = findFailures(output))
+    fails.isEmpty ||
+      let (msg = new Listing {
+        "Error processing \(if (fails.length > 1) "properties" else "property") in \(target):\n"
+        local maxPathLength = fails.map((f) -> f.path.join(pathSeparator).length).max
+        for (fail in fails) {
+          let (path = fail.path.join(pathSeparator))
+            "\t'\(path)'\(" ".repeat(maxPathLength - path.length)): \(fail.message)"
+        }
+      }.join("\n"))
+        throw(msg)
+
+local function longestPath(fails: List<Failure>) =
+  fails.map((f) -> f.path.join(pathSeparator).length).max
+
+local function findFailures(output: Mapping<String, Any?>|Dynamic): List<Failure> =
+  output.toMap().fold(List(), (acc: List<Failure>, _, value) ->
+    if (value is Failure)
+      acc.add(value)
+    else if (value is Mapping)
+      acc + findFailures(value)
+    else if (value is Dynamic)
+      acc + findFailures(value)
+    else
+      acc
+  )
```

Comment:
Speaking of functional patterns; if you write the accumulator as a callable function (i.e. give it a name), you don't need the distinction between gathering failures with `fold` and `+`.

```
local function findFailures(result: List<Failures>, output: Mapping<String, Any?>|Dynamic): List<Failures> =
  findFailuresAcc.apply(List(), null, output)

local findFailuresAcc = (result, _, value) ->
  if (value is Failure) acc.add(value)
  else if (value is Mapping|Dynamic) value.toMap().fold(acc, findFailuresAcc)
  else acc
```